### PR TITLE
expression: implement partial vectorized evaluation for `builtinCeilDecToDecSig`

### PR DESCRIPTION
**Pull request by TennyZhuang**
PCP #12102

### What problem does this PR solve? 

#12102

### What is changed and how it works?

vectorize `builtinCeilDecToDecSig`

```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinMathFunc/builtinCeilDecToDecSig-VecBuiltinFunc-4         	   16395	     67982 ns/op	   19488 B/op	     406 allocs/op
BenchmarkVectorizedBuiltinMathFunc/builtinCeilDecToDecSig-NonVecBuiltinFunc-4      	   10000	    116887 ns/op	   59664 B/op	    1243 allocs/op
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

 - vectorize `builtinCeilDecToDecSig`

### CODE REVIEW

File: expression/builtin_math_vec.go [hunk 1]
```diff
@@ -849,11 +849,47 @@ func (b *builtinTruncateUintSig) vecEvalInt(input *chunk.Chunk, result *chunk.Co
...
+		return err
+	}
+	result.ResizeDecimal(n, false)
+	result.MergeNulls(buf)
+	ds := result.Decimals()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		val := buf.GetDecimal(i)
```
Comment (SunRunAway) [inline expression/builtin_math_vec.go]: Decimal is a fexed-length value, consider using [`buf.Decimals()`](https://godoc.org/github.com/pingcap/tidb/util/chunk#Column.Decimals)

File: expression/builtin_math_vec.go [hunk 2]
```diff
@@ -849,11 +849,47 @@ func (b *builtinTruncateUintSig) vecEvalInt(input *chunk.Chunk, result *chunk.Co
...
+		var err error
+		if val.IsNegative() {
+			err = val.Round(&ds[i], 0, types.ModeTruncate)
```
Comment (SunRunAway) [inline expression/builtin_math_vec.go]: If `val` is negative, we should not add 1 below.

File: expression/builtin_math_vec.go [hunk 3]
```diff
@@ -849,11 +849,51 @@ func (b *builtinTruncateUintSig) vecEvalInt(input *chunk.Chunk, result *chunk.Co
...
+			continue
+		}
+		err = val.Round(&ds[i], 0, types.ModeTruncate)
+		if err != nil {
+			return err
+		}
+		if ds[i].Compare(val) == 0 {
+			continue
+		}
+		res := new(types.MyDecimal)
```
Comment (wshwsh12) [inline expression/builtin_math_vec.go]: In non-vec function, `val.Round(res, 0, types.ModeTruncate)` will set the `res` and it is not an empty MyDecimal now.

Comment (qw4990) [reply expression/builtin_math_vec.go]: @TennyZhuang Friendly ping, please address this comment.

File: expression/builtin_math_vec_test.go [hunk 4]
```diff
@@ -101,6 +101,7 @@ var vecBuiltinMathCases = map[string][]vecExprBenchCase{
 		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETInt}, childrenFieldTypes: []*types.FieldType{{Tp: mysql.TypeInt24}}, geners: nil},
 		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETDecimal}, geners: nil},
 		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETInt}, geners: nil},
+		{retEvalType: types.ETDecimal, childrenTypes: []types.EvalType{types.ETDecimal}, geners: nil},
```
Comment (francis0407) [inline expression/builtin_math_vec_test.go]: You have to define the `childrenFieldTypes` like what https://github.com/pingcap/tidb/pull/13409#discussion_r346683494 did.

**Comment by TennyZhuang** [comment]
PTAL @wshwsh12

**Comment by TennyZhuang** [comment]
I also update the benchmark result.

**Comment by TennyZhuang** [comment]
unit test failed due to 

```

[2019-11-18T11:12:03.780Z] 2019/11/18 19:12:03.635 terror.go:359: [error] [domain:8027]Information schema is out of date: schema failed to update in 1 lease, please make sure TiDB can connect to TiKV
[2019-11-18T11:12:03.780Z] github.com/pingcap/errors.AddStack
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/errors@v0.11.4/errors.go:174
[2019-11-18T11:12:03.780Z] github.com/pingcap/errors.Trace
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/errors@v0.11.4/juju_adaptor.go:15
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).checkSchemaValid
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1147
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).execute
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1092
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).executeAndWriteFinishBinlog
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1009
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/store/tikv.(*tikvTxn).Commit
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/store/tikv/txn.go:328
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*TxnState).Commit
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/txn.go:195
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).doCommit
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:440
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).doCommitWithRetry
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:459
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).CommitTxn
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:518
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.finishStmt
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/tidb.go:190
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.runStmt
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/tidb.go:272
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).executeStatement
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1019
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).execute
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1124
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/session.(*session).Execute
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/session/session.go:1059
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/ddl/util.CompleteDeleteRange
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/util/util.go:107
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/ddl.(*delRange).doTask
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/delete_range.go:223
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/ddl.(*delRange).doDelRangeWork
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/delete_range.go:177
[2019-11-18T11:12:03.780Z] github.com/pingcap/tidb/ddl.(*delRange).startEmulator
[2019-11-18T11:12:03.780Z] 	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/ddl/delete_range.go:141
[2019-11-18T11:12:03.780Z] runtime.goexit
[2019-11-18T11:12:03.780Z] 	/usr/local/go/src/runtime/asm_amd64.s:1357
```

It seems that not caused by this PR.

**Comment by you06** [comment]
No PCP points achieved because you submit this PR in the old team which is already closed now. Should I move this and others already completed PR to your new team?

**Pull request merged**